proposals/11-21-krm-fn-k8s-backend.md

```diff
@@ -0,0 +1,218 @@
+k8s doesn't allow to do this. That means that some krm-functions that require mounted files/directories
+just won't work with `--enable-as-a-pod` when they run from the client host. BUT if kustomize is ran from
+another pod that uses k8s volume, it will be possible to re-share the same volume that kustomize is using to
+the pod where krm-function is running. That means we have to give ability to user to configure a set of volumes for the pod.
+
+Keeping all that in mind now we can speak about the implementation options:
+
+The first option would be to give ability to user to simply set pod-label and list of volumes.
+This could be resolved by introducing of one more extra argument for labels and in reusing of the existing `--mount` argument.
+Or even with only one `--mount` argument, if we agree that the label will be always the constant `app: krm-pod`.
+We could select some default template for Pod we create, put there env vars, volumes and set the label and that implementation would work.
+
+Another approach is to allow the user to use the maximum set of possible options that k8s allows when Pod is created, e.g.,
+in addition to the existing volumes passed with arguments, the user should be able to provide ConfigMaps that also have to be
+mounted. That approach leads us to ability to create PodTemplate in advance and to use it to build krm-function pod.
+It will have a very similar outcome from security perspective with default values, but since it allows the user
+much more flexibility doing things in k8s-native way, this approach is selected to be implemented.
+It is implemented in [this PR](https://github.com/kubernetes-sigs/kustomize/pull/4260), but the current implementation
+doesn't allow to set the pod label via PodTemplate(TBD). To be able to do so we can either copy labels from PodTemplate, always set the as
+constant fields, or introduce one more extra argument. Since it's desired to minimize the number of arguments, the first option
+with copying labels from template will be selected. User may set the PodTemplate name with `--pod-template-name` argument and the pod
+template must exist in the same namespace where krm-function pod is going to be ran. The user may skip that parameter and by
+default the [following PodTemplate](https://github.com/kubernetes-sigs/kustomize/pull/4260/files#diff-bcc3694e89726203de4e763f2130891ec669833ab26f3856bcddde18c6514a9fR27)
+will be used. Env variables are put to that template already (TBD: maybe it's better to append env vars, rather than just replace the env vars already presented in the
+template). `--mount` flag is ignored as of now in case kubectl-backend is used, but there is a possible option to take what the set of volumes from PodTemplate
+and to append that with data from `--mount` flag - this is a potential topic for discussion.
+
+There is one more potential improvement that should be taken into account: with docker backend krm-function couldn't run
```

Comment:
Why a KRM fn may want to run another KRM fn? Do you have a concrete use case?

Reply:
Just as an example: At some point we had [kustomize_build krm function](https://github.com/GoogleContainerTools/kpt-functions-catalog/blob/959291b6d1275a21c2c9799f1ff9d92c3a40de55/functions/ts/src/kustomize_build.ts). It was [removed](https://github.com/GoogleContainerTools/kpt/issues/1078) because of several reasons, but let's say one will need to include some tool that runs krm-functions, e.g. kustomize, kpt into krm-function.. The `--pod-template-name` option is the only thing we need to support that case. I included that case mostly just to demonstrate that `pod-template` is a really flexible feature.

Reply:
included that example of krm function into the doc